### PR TITLE
Fix bugs and code improvements in Izhikevich neuron model

### PR DESCRIPTION
**Pull request by Hananel-Hazan**

### CODE REVIEW

File: bindsnet/network/nodes.py [hunk 1]
```diff
@@ -181,14 +181,14 @@ def step(self, inpts, dt):
         self.refrac_count[self.refrac_count != 0] -= dt
 
         # Check for spiking neurons.
-        self.s = (self.v >= self.thresh) & (self.refrac_count == 0)
```
Comment (djsaunde) [inline bindsnet/network/nodes.py]: This `& (self.refrac_count == 0)` is needed. It's not a bug. Neurons shouldn't spike if they're refractory.

File: bindsnet/network/nodes.py [hunk 2]
```diff
@@ -181,14 +181,14 @@ def step(self, inpts, dt):
...
+        self.s = (self.v >= self.thresh)
 
         # Refractoriness and voltage reset.
         self.refrac_count.masked_fill_(self.s, self.refrac)
         self.v.masked_fill_(self.s, self.reset)
 
         # Integrate input and decay voltages.
-        self.v += inpts
+        self.v += (self.refrac_count  == 0).float() * inpts
```
Comment (djsaunde) [inline bindsnet/network/nodes.py]: There's an extra space here: `(self.refrac_count  == 0).float() *` should be `(self.refrac_count == 0).float() *`

File: bindsnet/network/nodes.py [hunk 3]
```diff
@@ -249,7 +249,7 @@ def step(self, inpts, dt):
         self.refrac_count[self.refrac_count != 0] -= dt
 
         # Check for spiking neurons.
-        self.s = (self.v >= self.thresh) & (self.refrac_count == 0)
```
Comment (djsaunde) [inline bindsnet/network/nodes.py]: Same here.

File: bindsnet/network/nodes.py [hunk 4]
```diff
@@ -398,7 +398,7 @@ def step(self, inpts, dt):
         self.refrac_count[self.refrac_count != 0] -= dt
 
         # Check for spiking neurons.
-        self.s = (self.v >= self.thresh + self.theta) & (self.refrac_count == 0)
```
Comment (djsaunde) [inline bindsnet/network/nodes.py]: Here.

File: bindsnet/network/nodes.py [hunk 5]
```diff
@@ -473,7 +473,7 @@ def step(self, inpts, dt):
...
```
Comment (djsaunde) [inline bindsnet/network/nodes.py]: And here.

File: bindsnet/network/nodes.py [hunk 6]
```diff
@@ -525,11 +525,11 @@ def __init__(self, n=None, shape=None, traces=False, excitatory=True, rest=-65.0
         '''
         super().__init__(n, shape, traces, trace_tc)
 
-        self.rest = rest       # Rest voltage.
+        self.rest = rest         # Rest voltage.
```
Comment (djsaunde) [inline bindsnet/network/nodes.py]: Can you align these comments?

File: bindsnet/network/nodes.py [hunk 7]
```diff
@@ -557,18 +556,18 @@ def step(self, inpts, dt):
             | :code:`inpts` (:code:`torch.Tensor`): Inputs to the layer.
             | :code:`dt` (:code:`float`): Simulation time step.
         '''
-        # Decrement refrac counters.
-        self.refrac_count[self.refrac_count != 0] -= dt
 
         # Check for spiking neurons.
-        self.s = (self.v >= self.thresh) & (self.refrac_count == 0)
```
Comment (djsaunde) [inline bindsnet/network/nodes.py]: Here.

File: bindsnet/network/nodes.py [hunk 8]
```diff
@@ -557,18 +556,18 @@ def step(self, inpts, dt):
...
-        self.refrac_count.masked_fill_(self.s, self.refrac)
-        self.v.masked_fill_(self.s, self.reset)
+        self.v = torch.where(self.s, self.c, self.v)
+        self.u = torch.where(self.s, self.u + self.d, self.u)
+
 
         # Apply v and u updates.
-        self.v += dt * (0.04 * (self.v ** 2) + 5 * self.v + 140 - self.u + inpts)
+        self.v += dt * 0.5 * (0.04 * (self.v ** 2) + 5 * self.v + 140 - self.u + inpts)
+        self.v += dt * 0.5 * (0.04 * (self.v ** 2) + 5 * self.v + 140 - self.u + inpts)
```
Comment (djsaunde) [inline bindsnet/network/nodes.py]: This should only happen once.

File: bindsnet/network/nodes.py (hunk 1 above)
Comment (djsaunde) [reply bindsnet/network/nodes.py]: Actually, I see why this works now. Disregard these comments!

File: bindsnet/network/nodes.py (hunk 2 above)
Comment (Hananel-Hazan) [reply bindsnet/network/nodes.py]: I'm sorry, I dont see the different between the two snippets and I dont see the extra space

Comment (djsaunde) [reply bindsnet/network/nodes.py]: In your code, there's an extra space between `self.refrac_count` and the `==` operator.

File: bindsnet/network/nodes.py (hunk 8 above)
Comment (Hananel-Hazan) [reply bindsnet/network/nodes.py]: Our default dt = 1 ms, we getting the same computational instability that Izhikavitch has in the Matlab code

Comment (djsaunde) [reply bindsnet/network/nodes.py]: I anticipated this; that's why you should use `dt = 0.5`, instead of running the statement twice for all `dt`.

Comment (Hananel-Hazan) [reply bindsnet/network/nodes.py]: dt can only be integers.. otherwise we getting bugs...

Comment (djsaunde) [reply bindsnet/network/nodes.py]: No, `dt` is a `float`, or it should be, at least.

Comment (djsaunde) [reply bindsnet/network/nodes.py]: What bugs?

**Pull request merged**